programming_guide/section-2/section-2g/README.md

```diff
@@ -130,12 +130,83 @@ Waiting on DMAs associated with more than one object fifo:
+- **`mem`**: Reference to a host buffer, given as an argument to the sequence function, that this transfer will read from or write to. 
+- **`tensor_tile`** (optional): An alternative method to `offset`/`sizes`/`strides` for determining an access pattern over the `mem` buffer.
+- **`offset`** (optional): Starting point for the data transfer. Default values is `0`.
+- **`sizes`**: The extent of data to be transferred across each dimension. There is a maximum of four size dimensions.
+- **`strides`** (optional): Interval steps between data points in each dimension, useful for striding-across and reshaping data.
+- **`issue_token`** (optional): If a token is issued, one may call `dma_await_task` on the returned task. Default is `false`.
+
+The strides and strides express data transformations analogously to those described in [Section 2C](../section-2c).
+
+**Example Usage**:
+```python
+out_task = shim_dma_single_bd_task(of_out, C, sizes=[1, 1, 1, N], issue_token=True)
+```
+
+The example above describes a linear transfer of `N` data elements from the `C` buffer in host memory into an object FIFO with matching metadata labeled "of_out". The `sizes` dimensions are expressed right to left where the right is dimension 0 and the left dimension 3. Higher dimensions not used should be set to `1`.
+
+#### **Host Synchronization with `dma_await_task`**
+
+Synchronization between DMA channels and the host is facilitated by the `dma_await_task` operations, ensuring data consistency and proper execution order. The `dma_await_task` operation waits until the BD associated with the task is complete, issuing a task complete token.
+
+**Function Signature**:
+```python
+def dma_await_task(*args: DMAConfigureTaskForOp)
+```
+- **`args`: One or more `dma_task` objects, where `dma_task` objects are the value returned by `shim_dma_single_bd_task`.
+
+**Example Usage**:
+
+Waiting on DMAs associated with one object fifo:
+```python
+# Waits for the output data to transfer from the output object fifo to the host
+dma_await_task(out_task)  
+```
+
+Waiting on DMAs associated with more than one object fifo:
+```python
+dma_await_task(in_task, out_task)  
+```
+
+`dma_await_task` can only be called on a task created with `issue_token=True`. If `issue_token=False` (which is default), then `dma_free_task` should be called when the programmer knows that task if complete.
+
+#### **Best Practices for Data Movement and Synchronization with `dma_task` Operations**
+
+- **Await or Free to Reuse Buffer Descriptors**: While the exact buffer descriptor (BD) used for each operation is not visible to the user with the `dma_task` operations, there are still a finite number (maximum of `16` on a Shim Tile). Thus, it is important to use `dma_await_task` or `dma_free_task` before the number of BDs are exhausted so that they may be reused.
```

Comment:
Users may wonder when to use `free` vs. when to use `await` here. I suggest adding:

> `dma_free_task` allows the compiler to reuse the buffer descriptors of a task without synchronization. Using `dma_free_task(X)` before task `X` has completed will lead to a race condition and unpredictable behavior. Only use `dma_free_task(X)` in conjunction with some other means of synchronization. For example, you may issue `dma_free_task(X)` after a call to `dma_await_task(Y)` if you can reason that task `Y` can only complete _after_ task `X` has completed.

Reply:
I added a section with function signature and example uses for `dma_free_task` and also included your suggested explanation mostly verbatim.